src/tiled/editableobject.cpp

```diff
@@ -80,4 +85,90 @@ bool EditableObject::checkReadOnly() const
     return false;
 }
 
+static Map *mapForObject(Object *object)
+{
+    if (!object)
+        return nullptr;
+
+    switch (object->typeId()) {
+    case Object::LayerType:
+        return static_cast<Layer*>(object)->map();
+    case Object::MapObjectType:
+        return mapForObject(static_cast<MapObject*>(object)->objectGroup());
+    case Object::MapType:
+        return static_cast<Map*>(object);
+    case Object::ObjectTemplateType:
+    case Object::TerrainType:
+    case Object::TilesetType:
+    case Object::TileType:
+    case Object::WangSetType:
+    case Object::WangColorType:
+        break;
+    }
+    return nullptr;
+}
+
+QVariant EditableObject::toScript(const QVariant &value) const
+{
+    const int type = value.userType();
+
+    if (type == QMetaType::QVariantMap)
+        return toScript(value.toMap());
+
+    if (type == objectRefTypeId()) {
+        const auto ref = value.value<ObjectRef>();
+        MapObject *referencedObject = nullptr;
+
+        if (auto map = mapForObject(object())) {
+            referencedObject = map->findObjectById(ref.id);
+        } else if (object()->typeId() == Object::MapObjectType) {
+            if (auto objectGroup = static_cast<MapObject*>(object())->objectGroup()) {
```

Comment:
Do we want to support the case where there's a more complex detached layer tree? Currently, this just looks at the immediate parent of the map object. If a user does this:
```js
root = new GroupLayer();
child1 = new ObjectLayer();
child2 = new ObjectLayer();
root.addLayer(child1);
root.addLayer(child2);
obj1 = new MapObject();
obj2 = new MapObject();
child1.add(obj1);
child2.add(obj2);

// What happens here?
obj1.setProperty("test", obj2);
// Or here?
root.setProperty("test2", obj2);
```

I think we should fallback on just returning the id as an integer. The user can do any combination of detaching layers, attaching objects, and setting properties.

We should make a best-effort to locate the correct object by id; then return just the id if we can't find the object. This would only ever happen on detached layers / detached objects. JS is dynamically typed, so the user can check Number.isInteger() when working with it.
When we add a detached layer or object to a map (or tile), we should do something like (python-pseudocode):
```python
for object in added layers:
  for property in object.properties:
    if property.type == ObjectRef:
      # NOTE: Do this after adding the new layers, so that it can find id's inside the newly added layers.
      if object.asset() != asset() or not findObjectById(property.value.toInt()): 
        property.value = 0 # Unset
```

Reply:
I guess we'd have to add something like `object.originalAsset()` though.
And if `object.originalAsset()` is null, we should assume the object refs are intentional and try to resolve them.

Reply:
I think I only had tile object group in mind for this code, but indeed it also covers detached object groups. Sure we can make a best effort to handle detached hierarchies, and I like the idea of falling back to returning the `ObjectRef` itself (not the ID, because we need to keep the type of the value intact).

I don't think we should be actively invalidating object references though. As it is, you can also copy/paste them and as long as there is an object with a matching ID, it will refer to it. That behavior is somewhat predictable and I think preferable over the IDs getting set to 0. We just need to make sure it doesn't crash.

Reply:
That's reasonable.

What do you mean by returning the ObjectRef, not the ID? The object ref can't be manipulated by JS at all. Perhaps we need an `EditableObjectRef` class.

Reply:
We just need to do something similar to what I did with `FilePath` in change d13afac90f20271b81aeff00c56475512cf117c0.